server/services/src/test/java/org/zanata/adapter/TSAdapterTest.java

```diff
@@ -254,6 +258,74 @@ public void testFailToParseOriginalFile() throws Exception {
         }
     }
 
+    @Test
+    public void nonDocPartIsCaught() {
+        String base = "<TS version=\"2.1\" language=\"test\">";
+        Event event = new Event();
+        event.setEventType(EventType.START_DOCUMENT);
+        StartDocument part = new StartDocument();
+        part.setSkeleton(new GenericSkeleton(base));
+        event.setResource(part);
+        try {
+            adapter.replaceLocaleInDocPart(event, "en-GB");
+            fail("FileFormatAdapterException expected");
+        } catch (FileFormatAdapterException ffae) {
+            // Pass
+        }
+    }
+
+    @Test
+    public void badTargetLocaleIsCaught() {
+        String base = "<TS version=\"2.1\" language=\"test\">";
+        Event event = new Event();
+        event.setEventType(EventType.DOCUMENT_PART);
+        DocumentPart part = new DocumentPart();
+        part.setSkeleton(new GenericSkeleton(base));
+        event.setResource(part);
+        try {
+            adapter.replaceLocaleInDocPart(event, "$1");
+            fail("FileFormatAdapterException expected");
+        } catch (FileFormatAdapterException ffae) {
+            // Pass
+        }
+        try {
+            adapter.replaceLocaleInDocPart(event, "()");
+            fail("FileFormatAdapterException expected");
+        } catch (FileFormatAdapterException ffae) {
+            // Pass
+        }
+    }
+
+    @Test
+    public void validateLanguageReplacement() {
+        assertDocpartStringIsChanged("$badname", false);
+        assertDocpartStringIsChanged("bad=name", false);
+        assertDocpartStringIsChanged("badname=", false);
+        assertDocpartStringIsChanged("=badname", false);
+        assertDocpartStringIsChanged("bad name", false);
+        assertDocpartStringIsChanged("", true);
+        assertDocpartStringIsChanged("mo", true);
+        assertDocpartStringIsChanged("mo_PH", true);
+        assertDocpartStringIsChanged("mo-PH", true);
+        assertDocpartStringIsChanged("mo_PH.UTF-8", true);
+        assertDocpartStringIsChanged("mo_PH.UTF-8@5", true);
+    }
+
+    private void assertDocpartStringIsChanged(String originalLoc,
+                                              boolean changed) {
+        String base = "<TS version=\"2.1\" language=\"REPLACE\">";
```

Comment:
To protect against future breakage, could you please add another test case using the sourcelanguage attribute (which shouldn't be modified)?